### PR TITLE
kotlin: Move search context to the end of arguments

### PR DESCRIPTION
**Pull request by ebraminio**
This, in addition to annotating SearchContext interface as functional,
turns it into a more idiomatic Kotlin.

### CODE REVIEW

File: source/kotlin/src/test/kotlin/io/github/cosinekitty/astronomy/Tests.kt [hunk 1]
```diff
@@ -732,16 +732,11 @@ class Tests {
...
-            override fun eval(time: Time) = time.ut - cos(time.ut)
-        }
-
         val toleranceDays = 1.0e-9
         val toleranceSeconds = toleranceDays / 86400.0
-        val func = CosineIdentityFunc()
         val time1 = Time(0.0)
         val time2 = Time(1.0)
-        val tsolve = search(func, time1, time2, toleranceSeconds)
+        val tsolve = search(time1, time2, toleranceSeconds) { time -> time.ut - cos(time.ut) }
```
Comment (ebraminio) [inline source/kotlin/src/test/kotlin/io/github/cosinekitty/astronomy/Tests.kt:739]: When a functional interface object is put as the ending argument Kotlin can provide this syntax and that's why it is done always at the end.

See https://kotlinlang.org/docs/fun-interfaces.html for more information.

Comment (ebraminio) [reply source/kotlin/src/test/kotlin/io/github/cosinekitty/astronomy/Tests.kt:739]: https://kotlinlang.org/docs/lambdas.html#passing-trailing-lambdas

"According to Kotlin convention, if the last parameter of a function is a function, then a lambda expression passed as the corresponding argument can be placed outside the parentheses:"

**Comment by cosinekitty** [comment]
Wow, that is great!  When I started porting the `search` function, I kept debating whether I would pass it an interface or a lambda. I kept going back and forth in my mind, but settled on the interface approach because it is closer to what the other language versions do. (Over time, it is helpful to keep the versions somewhat similar, to make it easier when I have to fix bugs or add new features.) Now the "SAM" approach is very clever, because it allows using whichever approach makes more sense. I can go back and simplify much of the code now. Thank you!

**Comment by ebraminio** [comment]
SAM is essentially Java thing equivalent to C# delegates somehow and it will be better to use SAM so Java port can use that also as you did :)

**Pull request merged**